more_itertools/more.py

```diff
@@ -4521,3 +4522,40 @@ def partial_product(*iterables):
     for i, it in enumerate(iterators):
         for prod[i] in it:
             yield tuple(prod)
+
+
+def _discriminate_iterable(idx, iterable):
+    for item in iterable:
+        yield idx, item
+
+
+def full_outer_join(*iterables, key=lambda x: x):
+    """Lazily yields a full outer join of all iterables. The iterables are
+    expected to be sorted. Uses the merge join algorithm which runs in O(n)
+    time. Avoids buffering an entire iterable into memory.
+
+    To get an inner join of the iterables only consume the join keys that have
+    values for all iterables. To get a left/right/Nth join only consume the
+    join keys with values from your left/right/Nth iterable.
+
+    The default key function is an identity function.
+
+        >>> list(more_itertools.full_outer_join([{"id": 1, "val": "foo"}], [{"id": 1, "val": "bar"}, {"id": 2, "val": "baz"}], key=lambda x: x["id"]))
+        [(1, ([{'id': 1, 'val': 'foo'}], [{'id': 1, 'val': 'bar'}])), (2, ([], [{'id': 2, 'val': 'baz'}]))]
+    """
+    discriminated_iterables = [
+        _discriminate_iterable(iterable_idx, iterable)
+        for iterable_idx, iterable in enumerate(iterables)
+    ]
+
+    def _unwrapping_keyfunc(row):
+        return key(row[1])
+
+    for group_key, group in groupby(
+        merge(*discriminated_iterables, key=_unwrapping_keyfunc),
+        key=_unwrapping_keyfunc,
+    ):
+        key_batches = tuple(list() for _ in range(len(iterables)))
```

Comment:
Those N+2 function calls and N int lookups/creations are unnecessarily wasteful. Suggestion:
```python
key_batches = tuple([] for _ in iterables)
```